### ISSUE TITLE
precision and recall are very high, but the true positives (TP) in the confusion matrix are very low

### ISSUE THREAD

**Issue by tien0717** [ultralytics/ultralytics#14148]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

My precision and recall are very high, but the true positives (TP) in the confusion matrix are very low. Should I adjust the "iou: 0.7 # (float) intersection over union (IoU) threshold for NMS" or increase the number of background samples?

Here are my results and training parameters:
![results](https://github.com/ultralytics/ultralytics/assets/138637601/95390679-2e61-4d6e-ad1e-48416fc071e2)
![confusion_matrix_normalized](https://github.com/ultralytics/ultralytics/assets/138637601/8ba25f4d-02ba-4d8e-a780-71bb3b3626f1)
![F1_curve](https://github.com/ultralytics/ultralytics/assets/138637601/7e501474-eb61-4526-81a4-c9c78bacb0e7)
![P_curve](https://github.com/ultralytics/ultralytics/assets/138637601/c06b5218-30ea-4784-a3f8-31d4bc8dd4c6)

![PR_curve](https://github.com/ultralytics/ultralytics/assets/138637601/900064b7-d72a-4640-8e6a-f104d56888bc)
![R_curve](https://github.com/ultralytics/ultralytics/assets/138637601/f4ccaec7-0e91-40b3-b18c-ef7f10a88c5d)

task: detect
mode: train
model: yolov8n.yaml
data: crack.yaml
epochs: 200
time: null
patience: 100
batch: 8
imgsz: 640
save: true
save_period: 10
cache: false
device: null
workers: 0
project: null
name: train5
exist_ok: false
pretrained: true
optimizer: auto
verbose: true
seed: 0
deterministic: true
single_cls: false
rect: false
cos_lr: true
close_mosaic: 0
resume: false
amp: true
fraction: 1.0
profile: false
freeze: null
multi_scale: false
overlap_mask: true
mask_ratio: 4
dropout: 0.0
val: true
split: val
save_json: false
save_hybrid: false
conf: null
iou: 0.5
max_det: 300
half: false
dnn: false
plots: true
source: null
vid_stride: 1
stream_buffer: false
visualize: false
augment: false
agnostic_nms: false
classes: null
retina_masks: false
embed: null
show: false
save_frames: false
save_txt: false
save_conf: false
save_crop: false
show_labels: true
show_conf: true
show_boxes: true
line_width: null
format: torchscript
keras: false
optimize: false
int8: false
dynamic: false
simplify: false
opset: null
workspace: 4
nms: false
lr0: 0.001
lrf: 0.0005
momentum: 0.937
weight_decay: 0.0005
warmup_epochs: 3.0
warmup_momentum: 0.8
warmup_bias_lr: 0.1
box: 7.5
cls: 0.5
dfl: 1.5
pose: 12.0
kobj: 1.0
label_smoothing: 0.0
nbs: 64
hsv_h: 0.015
hsv_s: 0.7
hsv_v: 0.4
degrees: 0.0
translate: 0.1
scale: 0.5
shear: 0.01
perspective: 0
flipud: 0.0
fliplr: 0.5
bgr: 0.0
mosaic: 0.5
mixup: 0.1
copy_paste: 0.05
auto_augment: randaugment
erasing: 0.4
crop_fraction: 1.0
cfg: null
tracker: botsort.yaml
save_dir: runs\detect\train5

**Comment by Y-T-G** [comment]
They aren't high. The confusion matrix is calculated at 0.45 conf threshold. It's hardcoded. Your recall at 0.45 is less that 0.2

**Comment by tien0717** [comment]
@Y-T-G how can I improve the true positives (TP) in the confusion matrix?

**Comment by glenn-jocher** [comment]
Thank you for your detailed question and for sharing your results and training parameters. To improve the true positives (TP) in your confusion matrix, here are a few suggestions:

1. **Model Variant**: As mentioned, using a larger model variant like YOLOv8m can help improve detection accuracy, especially for more complex tasks. Larger models have more capacity to learn intricate patterns in the data.

2. **IoU Threshold**: Adjusting the IoU threshold for Non-Maximum Suppression (NMS) can also help. You can experiment with different IoU values to see which one yields better results for your specific dataset. For instance, you might try lowering the IoU threshold to 0.5 or 0.6 to see if it improves the TP rate.

3. **Data Augmentation**: Enhancing your data augmentation techniques can help the model generalize better. You can try increasing the mosaic, mixup, or other augmentation parameters to introduce more variability in the training data.

4. **Training Parameters**: Fine-tuning your training parameters such as learning rate, batch size, and number of epochs can also make a significant difference. Ensure that your learning rate is appropriate for your dataset and that you are training for enough epochs to allow the model to converge.

5. **Class Imbalance**: If your dataset has class imbalance, consider techniques like oversampling the minority class or using class weights to ensure the model does not become biased towards the majority class.

6. **Additional Resources**: For more detailed strategies on improving model accuracy, especially for small objects, you can refer to the tutorial linked: [YOLOv8 Accuracy Improvement](https://y-t-g.github.io/tutorials/yolov8-increase-accuracy/).

Here’s a quick example of how you might adjust the IoU threshold in your training script:

```python
from ultralytics import YOLO

# Load your model
model = YOLO('yolov8m.yaml')  # or yolov8m.pt for a pretrained model

# Train the model with adjusted parameters
model.train(data='crack.yaml', epochs=200, batch=8, iou=0.5, lr0=0.001)
```

Feel free to experiment with these suggestions and monitor the impact on your TP rate. If you continue to face issues, providing a minimum reproducible example can help us diagnose the problem more effectively. You can find more information on creating a reproducible example here: [Minimum Reproducible Example](https://docs.ultralytics.com/help/minimum_reproducible_example).

Best of luck with your training, and feel free to reach out if you have any more questions! 😊
